packages/core/core/src/RequestGraph.js

```diff
@@ -0,0 +1,227 @@
+// @flow strict-local
+import {PromiseQueue, md5FromString} from '@parcel/utils';
+import type {
+  AssetRequest,
+  Config,
+  FilePath,
+  ParcelOptions
+} from '@parcel/types';
+import type {Event} from '@parcel/watcher';
+import WorkerFarm from '@parcel/workers';
+
+import Dependency from './Dependency';
+import Graph, {type GraphOpts} from './Graph';
+import ResolverRunner from './ResolverRunner';
+import type {
+  AssetRequestNode,
+  CacheEntry,
+  DepPathRequestNode,
+  NodeId,
+  RequestGraphNode,
+  RequestNode,
+  RequestResult
+} from './types';
+
+type RequestGraphOpts = {|
+  ...GraphOpts<RequestGraphNode>,
+  config: Config,
+  options: ParcelOptions,
+  onAssetRequestComplete: (AssetRequestNode, CacheEntry) => mixed,
+  onDepPathRequestComplete: (DepPathRequestNode, AssetRequest | null) => mixed
+|};
+
+const hashObject = obj => {
+  return md5FromString(JSON.stringify(obj));
+};
+
+const nodeFromDepPathRequest = (dep: Dependency) => ({
+  id: dep.id,
+  type: 'dep_path_request',
+  value: dep
+});
+
+const nodeFromAssetRequest = (assetRequest: AssetRequest) => ({
+  id: hashObject(assetRequest),
+  type: 'asset_request',
+  value: assetRequest
+});
+
+const nodeFromFilePath = (filePath: string) => ({
+  id: filePath,
+  type: 'file',
+  value: {filePath}
+});
+
+export default class RequestGraph extends Graph<RequestGraphNode> {
+  inProgress: Map<NodeId, Promise<RequestResult>> = new Map();
+  invalidNodes: Map<NodeId, RequestNode> = new Map();
+  runTransform: (file: AssetRequest) => Promise<CacheEntry>;
+  resolverRunner: ResolverRunner;
+  onAssetRequestComplete: (AssetRequestNode, CacheEntry) => mixed;
+  onDepPathRequestComplete: (DepPathRequestNode, AssetRequest | null) => mixed;
+  queue: PromiseQueue;
+  farm: WorkerFarm;
+
+  constructor({
+    onAssetRequestComplete,
+    onDepPathRequestComplete,
+    config,
+    options,
+    ...graphOpts
+  }: RequestGraphOpts) {
+    super(graphOpts);
+    this.queue = new PromiseQueue();
+    this.onAssetRequestComplete = onAssetRequestComplete;
+    this.onDepPathRequestComplete = onDepPathRequestComplete;
+
+    this.resolverRunner = new ResolverRunner({
+      config,
+      options
+    });
+  }
+
+  async initFarm() {
+    // This expects the worker farm to already be initialized by Parcel prior to calling
+    // AssetGraphBuilder, which avoids needing to pass the options through here.
+    this.farm = await WorkerFarm.getShared();
+    this.runTransform = this.farm.mkhandle('runTransform');
```

Comment:
I wonder about having the graph being the one doing the actual resolving and transforming, vs being a data structure...

Reply:
@wbinnssmith and I were talking about making both the `RequestGraph` and `AssetGraph` have instances of a `Graph`. Mainly we were talking about not exposing low level graph methods to users of `AssetGraph`, which is less of a concern for the `RequestGraph` since it is only used by `AssetGraphBuilder`, but that would take care of any concern of having the data structure do async stuff. I like the idea, but I think we can address that in a future PR.